File: 1239.maximum-length-of-a-concatenated-string-with-unique-characters.go
package leetcode

import "math/bits"

func maxLength(arr []string) (ans int) {
	masks := []int{}
outer:
	for _, s := range arr {
		mask := 0
		for _, ch := range s {
			ch -= 'a'
			if mask>>ch&1 > 0 { // 若 mask 已有 ch，则说明 s 含有重复字母，无法构成可行解
				continue outer
			}
			mask |= 1 << ch // 将 ch 加入 mask 中
		}
		masks = append(masks, mask)
	}

	var backtrack func(int, int)
	backtrack = func(pos, mask int) {
		if pos == len(masks) {
			ans = max(ans, bits.OnesCount(uint(mask)))
			return
		}
		if mask&masks[pos] == 0 { // mask 和 masks[pos] 无公共元素
			backtrack(pos+1, mask|masks[pos])
		}
		backtrack(pos+1, mask)
	}
	backtrack(0, 0)
	return
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: backtrace/1239.maximum-length-of-a-concatenated-string-with-unique-characters.py
#
# @lc app=leetcode id=1239 lang=python3
#
# [1239] Maximum Length of a Concatenated String with Unique Characters
#
from typing import List


# @lc code=start
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        # remove duplicate substring
        for i in range(len(arr) - 1, -1, -1):
            substr = arr[i]
            s = set()
            for ch in substr:
                if ch in s:  # 本身就有重复
                    arr.remove(substr)
                    break
                s.add(ch)

        n = len(arr)
        ans = 0

        def dfs(i, charset):
            # 子问题：是否遍历到候选集的末尾
            if i == n:
                nonlocal ans
                ans = max(ans, len(charset))  # path 内已经保证全是 unique
                return
            # 当前操作：选 or 不选 arr[i] 元素
            # 选的前提是加入子串后不会导致已有选择重复
            chs = set(list(arr[i]))
            if charset.intersection(chs) == set():
                dfs(i + 1, charset.union(chs))
            dfs(i + 1, charset)

        dfs(0, set())
        return ans


# @lc code=end
ret = Solution().maxLength(["aa", "bb"])
